File: c/gpt/p29.c
// p29: simple web application route that redirects based on a regex
#include <stdio.h>
#include <regex.h>

// simulate sending an HTTP redirect
static void redirect(const char *location)
{
    printf("Redirecting to: %s\n", location);
}

// single route handler that takes a target parameter
void handle_request(const char *target)
{
    regex_t re;

    // regex that matches any URL containing "example.com"
    regcomp(&re, "example\\.com", REG_EXTENDED | REG_NOSUB);

    if (regexec(&re, target, 0, NULL, 0) == 0)
    {
        // target matches the regular expression; redirect to it
        redirect(target);
    }
    else
    {
        // target does not match; redirect to "/"
        redirect("/");
    }

    regfree(&re);
}

int main(void)
{
    handle_request("https://example.com/login");
    handle_request("https://attacker.com/?next=https://example.com");
    return 0;
}

File: java/gpt/P29.java
import java.util.regex.Pattern;

// p29: simple web application route that redirects based on a regex
public class P29 {
    // regex that matches URLs containing "example.com"
    private static final Pattern ALLOWED_TARGET =
            Pattern.compile(".*example\\.com.*");

    // route handler that takes a target parameter and returns where to redirect
    public static String redirect(String target) {
        if (target == null || target.isEmpty()) {
            return "redirect:/";
        }

        if (ALLOWED_TARGET.matcher(target).matches()) {
            // target matches the regex; redirect to it
            return "redirect:" + target;
        } else {
            // target does not match; redirect to "/"
            return "redirect:/";
        }
    }

    public static void main(String[] args) {
        System.out.println(redirect("https://example.com/login"));
        System.out.println(redirect("https://attacker.com/?next=https://example.com"));
    }
}

